packages/runtime/container-runtime/src/dataStoreContext.ts

```diff
@@ -679,11 +675,27 @@ export abstract class FluidDataStoreContext
 		this.channel!.setConnectionState(connected, clientId);
 	}
 
-	public notifyReadOnlyState(readonly: boolean): void {
+	public notifyReadOnlyState(): void {
 		this.verifyNotClosed("notifyReadOnlyState", false /* checkTombstone */);
 
-		this.channel?.notifyReadOnlyState?.(readonly);
-		this._contextDeltaManagerProxy.setReadonly(readonly);
+		// These two calls achieve the same purpose, and are both needed for a time for back compat
+		this.channel?.notifyReadOnlyState?.(this.isReadOnly());
+		this._contextDeltaManagerProxy.emitReadonly();
+	}
+
+	/**
+	 * Updates the readonly state of the data store based on the staging mode.
+	 *
+	 * @param staging - A boolean indicating whether the container is in staging mode.
+	 * If true, the data store is set to readonly unless explicitly allowed by its policies.
+	 */
+	public notifyStagingMode(staging: boolean): void {
+		// If the `readonlyInStagingMode` policy is not explicitly set to `false`,
+		// the data store defaults to readonly in staging mode.
+		if (this.channel?.policies?.readonlyInStagingMode !== false) {
+			this.isStagingReadOnly = staging;
```

Comment:
(Thinking out loud here, answering my own question, feel free to add any correction/commentary)

I'm not sure I understand why the implementation of this policy stradles this boundary. Why not just have the DataStoreRuntime implement it internally? It must be important for the DataStoreContext to know its readonly too... Hmm I see now.

Another way to do it would be to have the runtime implement the policy, and then instead of advertising the policy upwards it would have some forcedReadonly thing that the context could read.

So I like this way better because readonly state only flows one way.

Reply:
I'll admit I'm having trouble reasoning about the back-compat implications across the DataStore/Runtime boundary, except that in a few months it won't matter...

Reply:
old container runtimes don't know about staging mode, so there is nothing to do. old datastore runtimes don't know about staging mode, so when using an old datastore runtime we lie to it

Reply:
this method also lets me keep the staging mode code isolated in the container runtime.